synfig-studio/src/gui/workarearenderer/renderer_canvas.cpp

```diff
@@ -371,19 +376,70 @@ Renderer_Canvas::build_onion_frames()
 		const Color color_future(0.f, 1.f, 0.f, 0.2f);
 		const ColorReal base_alpha = 1.f;
 		const ColorReal current_alpha = 0.5f;
-		// make onion levels
-		for(int i = past; i > 0; --i) {
-			Time time = base_time - frame_duration*i;
-			ColorReal alpha = base_alpha + (ColorReal)(past - i + 1)/(ColorReal)(past + 1);
-			if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
-				onion_frames.push_back(FrameDesc(time, w, h, alpha));
+
+		if (keyframes) {
+			// Onion Skin on Keyframes
+			// Find and push_back valid Past Keyframes Times
+			Time keyframe_time = base_time;
+			for(int i = past; i > 0; --i) {
+				KeyframeList::iterator iter;
+				if (canvas->keyframe_list().find_prev(keyframe_time, iter)) {
+					if (keyframe_time >= rend_desc.get_time_start() && keyframe_time <= rend_desc.get_time_end()) {
+						keyframe_time = iter->get_time();
+						past_times.push_back(keyframe_time);
+					}
+				}
+				else
+					break;
+			}
+
+			// Find and push_back valid Future Keyframes Times
+			keyframe_time = base_time;
+			for(int i = future; i > 0; --i) {
+				KeyframeList::iterator iter;
+				if (canvas->keyframe_list().find_next(keyframe_time, iter)) {
+					if (keyframe_time >= rend_desc.get_time_start() && keyframe_time <= rend_desc.get_time_end()) {
+						keyframe_time = iter->get_time();
+						future_times.push_back(keyframe_time);
+					}
+				}
+				else
+					break;
+			}
+		} else {
+			// Onion Skin on Frames
+			// Find and push_back valid Past Frames Times
+			for(int i = past; i > 0; --i) {
+				Time frame_time = base_time - frame_duration * i;
+				if (frame_time >= rend_desc.get_time_start() && frame_time <= rend_desc.get_time_end())
+					past_times.push_back(frame_time);
+			}
+
+			// Find and push_back valid Future Frames Times
+			for(int i = future; i > 0; --i) {
+				Time frame_time = base_time + frame_duration * i;
+				if (frame_time >= rend_desc.get_time_start() && frame_time <= rend_desc.get_time_end())
+					future_times.push_back(frame_time);
+			}
 		}
-		for(int i = future; i > 0; --i) {
-			Time time = base_time + frame_duration*i;
-			ColorReal alpha = base_alpha + (ColorReal)(future - i + 1)/(ColorReal)(future + 1);
-			if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
-				onion_frames.push_back(FrameDesc(time, w, h, alpha));
+
+		// Check number of available Past and Future Frames
+		past_frames = (past > past_times.size()) ? past_times.size() : past;
+		future_frames = (future > future_times.size()) ? future_times.size() : future;
```

Comment:
They are only used since this point. So declare them here, not up there.

And, actually, there is no need to check the values anymore, they can be initialized directly the vector sizes. They are already limited by `past`/`future` before in their loops.

Reply:
Completed!